### PR TITLE
Add Yarn support in new apps using --yarn option

### PR DESCRIPTION
**Pull request by Liceth**
This add a package.json and the settings needed to get npm nodules integrated in new apps when the --yarn option is used (e.g `rails new blog --yarn`).

The next changes should be done to finish this:
- [x] Add `package.json` in new apps when `--yarn` option is passed
- [x] Run `yarn install` after `bundle install` in new apps
- [x] Add rails-ujs dependency to `package.json`
- [x] Add `node_modules` folder to asset paths
- [x] Fix existing broken tests
- [x] Add new tests
- [x] Changelog

### CODE REVIEW

File: railties/lib/rails/generators/app_base.rb [hunk 1]
```diff
@@ -404,6 +412,49 @@ def run_bundle
         bundle_command("install") if bundle_install?
       end
 
+      def npm_path
+        commands = ["npm"]
```
Comment (guilleiguaran) [inline railties/lib/rails/generators/app_base.rb]: I think we can add also to "yarn" command here (https://code.facebook.com/posts/1840075619545360)

File: railties/lib/rails/generators/app_base.rb [hunk 2]
```diff
@@ -33,6 +33,9 @@ def self.add_shared_options_for(name)
         class_option :javascript,         type: :string, aliases: "-j", default: "jquery",
                                           desc: "Preconfigure for selected JavaScript library"
 
+        class_option :npm,                type: :boolean, default: false,
+                                          desc: "Preconfigure for NPM for assets management"
+
```
Comment (guilleiguaran) [inline railties/lib/rails/generators/app_base.rb]: "Preconfigure NPM for assets management"

File: railties/lib/rails/generators/rails/app/templates/package.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "name": "<%= app_name %>",
+  "private": true,
+  "dependencies": {
+    "jquery": "~3.1.0",
+    "jquery-ujs": "~1.2.2"
+  }
```
Comment (guilleiguaran) [inline railties/lib/rails/generators/rails/app/templates/package.json]: I think we will be able to remove jquery from here and change `jquery-ujs` to `rails-ujs` after of having https://github.com/rails/rails/issues/25208 finished

File: railties/lib/rails/generators/app_base.rb [hunk 4]
```diff
@@ -404,6 +412,49 @@ def run_bundle
...
+          path = paths.find do |p|
+            full_path = File.expand_path(cmd, p)
+            File.executable?(full_path) && File.file?(full_path)
+          end
+          path && File.expand_path(cmd, path)
+        end
+
+        unless npm_path
+          puts %{Warning: npm option passed but npm executable was 
+            not detected in the system. Download Node.js in https://nodejs.org/en/download/}
```
Comment (guilleiguaran) [inline railties/lib/rails/generators/app_base.rb]: Let's move this warning message to https://github.com/rails/rails/pull/26836/files#diff-248322211d56bb3b2ab59ac68a2839f5R212, right now it's being displayed everytime the npm_path function is called if command is not found

**Comment by rafaelfranca** [comment]
Can we change npm with yarn? I think only the npm install command that
needs to change

**Comment by Liceth** [comment]
yes, we can add "yarn" as first option to supported commands list and it will be used instead of npm if installed

**Comment by rafaelfranca** [comment]
I'd not even bother to support npm, just jump straight to yarn.

On Wed, 19 Oct 2016 at 22:57 Liceth Ovalles Rodriguez <
notifications@github.com> wrote:

yes, we can add "yarn" as first option to supported commands list and it
will be used instead of npm if installed

—
You are receiving this because you commented.


Reply to this email directly, view it on GitHub
<https://github.com/rails/rails/pull/26836#issuecomment-254989318>, or mute
the thread

**Comment by Liceth** [comment]
@eileencodes @rafaelfranca @guilleiguaran This is ready for review and failures in TravisCI aren't related to my changes.

**Comment by guilleiguaran** [comment]
@Liceth I've fixed the mysql2 tests in f13ec72664fd13d3 and updated your branch

File: railties/lib/rails/generators/rails/app/app_generator.rb [hunk 5]
```diff
@@ -205,6 +209,9 @@ def create_root_files
         build(:readme)
         build(:rakefile)
         build(:configru)
+        if options[:npm]
+          build(:packagejson)
+        end
```
Comment (guilleiguaran) [inline railties/lib/rails/generators/rails/app/app_generator.rb]: This if can be done in a single line:

```build(:packagejson) if options[:npm]```

Comment (Liceth) [inline railties/lib/rails/generators/rails/app/app_generator.rb]: Thanks for the observation @guilleiguaran I already made the changes

**Comment by bouk** [comment]
Please note that yarn is not a total drop-in replacement for npm, we can't try one or the other. A project that uses yarn gets a yarn.lock file (analogous to Gemfile.lock). NPM doesn't create a file like that and doesn't interpret it.

I fully support not even bothering with NPM and just using yarn, it's a lot more robust in my experience. Requiring people to install it separately might take away from the plug-and-play nature of Rails.

**Comment by Liceth** [comment]
I'm fine where ever you decide about yarn or npm, only let me know the finally decision on it.

**Comment by connorshea** [comment]
Don't you install Yarn using npm? I 100% support using Yarn as the package manager, but it won't avoid npm entirely.

File: railties/lib/rails/generators/app_base.rb [hunk 6]
```diff
@@ -33,6 +33,9 @@ def self.add_shared_options_for(name)
         class_option :javascript,         type: :string, aliases: "-j", default: "jquery",
                                           desc: "Preconfigure for selected JavaScript library"
 
+        class_option :npm,                type: :boolean, default: false,
+                                          desc: "Preconfigure for NPM assets management"
```
Comment (connorshea) [inline railties/lib/rails/generators/app_base.rb]: npm should be lowercase in all situations as far as I know.

File: railties/lib/rails/generators/app_base.rb [hunk 7]
```diff
@@ -400,10 +406,55 @@ def os_supports_listen_out_of_the_box?
...
+      end
+
+      def run_npm
+        if package_json_exist?
+          if npm_path
+            say_status :run, "npm install"
+            npm_command("install")
+          else
+            say_status :warning, "npm option passed but npm executable was not detected in the system.", :yellow
+            say_status :warning, "Download Node.js in https://nodejs.org/en/download/.", :yellow
```
Comment (connorshea) [inline railties/lib/rails/generators/app_base.rb]: Should be "Download Node.js at https://nodejs.org/en/download/"

Comment (Liceth) [inline railties/lib/rails/generators/app_base.rb]: Hi @connorshea Thanks for you review

**Comment by bouk** [comment]
For installing yarn there's options available for most common package managers (brew, apt, Chocolatey) https://yarnpkg.com/en/docs/install

**Comment by bouk** [comment]
It might be handy to add updating of the JS packages to the `bin/update` and `bin/setup` templates

**Comment by guilleiguaran** [comment]
@Liceth we have decided to skip NPM and go directly with Yarn because of the improvements of this last. Can you update your pull request? 

Also update the `bin/update` and `bin/setup` files as @bouk suggested.

Thank you very much for your work on this!!!

**Comment by Liceth** [comment]
Sure @guilleiguaran  I wil do the changes in the pull request.
 Thank you 👍

**Comment by Liceth** [comment]
@rafaelfranca @guilleiguaran I'm finishing this but I have a quick question related to jQuery removal (#25208)

If the users pass the -j option when they're creating a new app do we need to add the library to the Gemfile (e.g jquery-rails) or to the package.json?

**Comment by guilleiguaran** [comment]
Let's keep the actual behavior of -j option, if the users pass it let's add the gem instead of the npm dependency.

**Comment by guilleiguaran** [comment]
Also please rebase this and squash commits

**Comment by guilleiguaran** [comment]
This seems ready for me, I've assigned to @dhh for final review

**Comment by guilleiguaran** [comment]
One last thing, please add a CHANGELOG entry for this, please!!

**Comment by Liceth** [comment]
Ok I will do the change @guilleiguaran

**Comment by guilleiguaran** [comment]
Sorry, I was supposed to unassign myself only, not to @rafaelfranca,  re-assigned to him also 😄

**Comment by dhh** [comment]
What options do we have for relocating this config file somewhere else? IMO, it seems natural that this should either live in config/ or app/assets. But not in root.

**Comment by guilleiguaran** [comment]
AFAIK all the operations done with the `yarn` command are relative to the current folder (used as prefix), so it we want to move `package.json` to `app/assets` we need to move also `node_modules` and users will need to run `yarn` commands inside of `app/assets` folder.

File: railties/lib/rails/generators/app_base.rb [hunk 8]
```diff
@@ -33,6 +33,9 @@ def self.add_shared_options_for(name)
         class_option :javascript,         type: :string, aliases: "-j",
                                           desc: "Preconfigure for selected JavaScript library"
 
+        class_option :yarn,                type: :boolean, default: false,
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/app_base.rb]: wrong indentation here.

File: railties/lib/rails/generators/app_base.rb [hunk 9]
```diff
@@ -410,10 +413,55 @@ def os_supports_listen_out_of_the_box?
...
+
+        if RbConfig::CONFIG["host_os"] =~ /mswin|cygwin/
+          ENV["PATHEXT"].split(File::PATH_SEPARATOR).each do |ext|
+            commands << commands[0] + ext
+          end
+        end
+
+        yarn_path = commands.find do |cmd|
+          paths = ENV["PATH"].split(File::PATH_SEPARATOR)
+          path = paths.find do |p|
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/app_base.rb]: blank line around this block

File: railties/lib/rails/generators/app_base.rb [hunk 10]
```diff
@@ -410,10 +413,55 @@ def os_supports_listen_out_of_the_box?
...
+            full_path = File.expand_path(cmd, p)
+            File.executable?(full_path) && File.file?(full_path)
+          end
+          path && File.expand_path(cmd, path)
+        end
+        yarn_path
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/app_base.rb]: blank line before this line

File: railties/lib/rails/generators/app_base.rb [hunk 11]
```diff
@@ -410,10 +413,55 @@ def os_supports_listen_out_of_the_box?
...
+      end
+
+      def yarn_command(command)
+        full_command = "#{yarn_path} #{command}"
+        if options[:quiet]
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/app_base.rb]: blank line before this line

File: railties/lib/rails/generators/app_base.rb [hunk 12]
```diff
@@ -410,10 +413,55 @@ def os_supports_listen_out_of_the_box?
...
+          system(full_command, out: File::NULL)
+        else
+          system(full_command)
+        end
+      end
+
+      def run_yarn
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/app_base.rb]: This method should be the first defined and the others be defined after this one.

```
def run_yean
def package_json_exist?
def yarn_path
def yarn_command
```

File: railties/lib/rails/generators/rails/app/templates/bin/setup.tt [hunk 13]
```diff
@@ -16,8 +16,11 @@ chdir APP_ROOT do
   puts '== Installing dependencies =='
   system! 'gem install bundler --conservative'
   system('bundle check') || system!('bundle install')
-<% unless options.skip_active_record -%>
+<% if options[:yarn] %>
+  system! 'yarn'
+<% end %>
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/rails/app/templates/bin/setup.tt]: Should not the blank line be inside the conditional?

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

File: railties/lib/rails/generators/rails/app/templates/bin/setup.tt (hunk 13 above)
Comment (guilleiguaran) [inline railties/lib/rails/generators/rails/app/templates/bin/setup.tt]: 👍

**Comment by Liceth** [comment]
@rafaelfranca Changes are done

**Review by rafaelfranca — APPROVED** [review]
Thank you so much.

Please add a CHANGELOG entry so @guilleiguaran can merge it.

**Comment by guilleiguaran** [comment]
Looks like the CHANGELOG entry was missed with the last rebase but I still had it locally so I've pushed it in 3b0a09c398ba9

**Comment by Liceth** [comment]
Thank you!!  👍

**Comment by vipulnsward** [comment]
🚀 ❤️

**Comment by dhh** [comment]
Thanks for working on this @Liceth! I've been adding Yarn to Basecamp 3 and made a few alterations to the setup merged here that I'd like to see in master, if you're interested in continuing to work on this:

* Having node_modules in the root of the app doesn't feel quite right. I get that this is the standard for full-JS apps, but a Rails app is more like a cohabitation space. So I'd like for us to use `vendor/yarn/node_modules` (and `vendor/yarn/package.json`).
* To use a custom placement for Yarn, we'll need to make sure Yarn is run from `vendor/yarn`. To that effect we can use a new `bin/yarn` binstub that sets the things right. I've been running with this binstub: 
```
#!/bin/bash
app_root="$( cd "$(dirname "$0")/.."; pwd )"
( cd $app_root/vendor/yarn; yarn $@ )
```
* We should have a `vendor/yarn/.gitignore` that has `node_modules` in it.
* I think we should also have a `vendor/assets/javascripts/yarn.js` stub that suggests how to use `//= require some/yarned/js/stuff` to be compiled by the asset pipeline. We can then include this yarn dependency file from the default application.js setup.
* I'd like to consider making Yarn the default when its installed on the system. So if `yarn` is available, we switch to Yarn-by-default and expose `--skip-yarn` instead.

**Comment by dhh** [comment]
(BTW, I'm aware that the default node load path scheme is easier when you nest node_modules in a clear hierarchy. With `vendor/yarn/node_modules` AND `app/assets/javascripts` we'll need to set NODE_PATH or similar when configuring web pack. But I think that's fine.)

**Comment by keoghpe** [comment]
:+1: Great job, I feel rails has been _yarning_ for this for quite some time.

**Comment by bouk** [comment]
@dhh I think your suggestion makes sense as a default for Rails, and apps can change their set up if needs arise

**Comment by dhh** [comment]
On second thought, perhaps we don't even need the yarn directory. We could just go with `vendor/node_modules`, `vendor/package.json`, `vendor/yarn.lock`. That actually seems well enough.

**Comment by benlangfeld** [comment]
@dhh If package.json has to go in vendor/, then Gemfile logically does too. I don't object to node_modules being hidden away, but package.json are root level application concerns and should be exposed just like Gemfile, .travis.yml, etc

**Comment by dhh** [comment]
I think there's a great argument for Gemfile going in vendor too. But that requires moving something we have to that. Definitely something to look into but separate from compounding the problem.

And it's not just package.json, it's also the Yarn lock file.

**Comment by dhh** [comment]
Implemented my recommendations for using vendor in https://github.com/rails/rails/pull/27245.

**Comment by benlangfeld** [comment]
@dhh I still don't understand the objection to leaving those project metadata files in the root directory where all the others like it live. What am I missing?

**Comment by dhh** [comment]
@benlangfeld The root is enough of a mess as-is in many projects. To clean that up, we gotta start somewhere. This seems like a good place. package.json/yarn.lock and the node_modules directory are all clearly vendor concerns. So moving that stuff to vendor makes sense to me.

**Comment by dhh** [comment]
I'd be happy in time to move Gemfile/Gemfile.lock there as well.

**Comment by benlangfeld** [comment]
@dhh In cases where Rails engines are also npm modules for the purpose of delivering assets, the package.json they have is meta-data about the package being delivered in the same vain as the gemspec. In the event this change propagated to engines, would you support these files staying in the engine root alongside the gemspec?

I ask mostly because this is something I'm working through on our engine-based app, but also because I recall a traumatic incident as a child when I "tidied up" some messy DLLs from C:/ to C:/Windows and "bricked" the family computer. 😄

**Comment by dhh** [comment]
Can you walk me through the workflow you have for engines where there is a package.json? Are you using that file to publish the JS parts of the engine to NPM? Or are you thinking that this is a way to integrate matters once we get to Webpack?

Because I get that package.json is basically both Gemfile and .gemspec in one. In Rails, your own app doesn't have a .gemspec, but your engine might. Here, we're using package.json as Gemfile, just to manage the vendored JS dependencies, not as .gemspec with the intention of publishing the app somewhere.

**Comment by benlangfeld** [comment]
@dhh We're following an internal workflow based on the ideas at http://cbra.info, the overall goal being rescuing a 200kloc, 600 model application from chaos.

This app also uses [react_on_rails](https://github.com/shakacode/react_on_rails), which uses Webpack rather than the asset pipeline, and my current effort is on linking engine's assets into the app. This is what I'm actively working on as of today, so my ideas are not yet well formed.

> once we get to Webpack?

Is some core usage of Webpack already on the roadmap? Are there details available of what this will look like?

**Comment by dhh** [comment]
Gotcha. Yeah, let's try to sort it all out.

I'm starting work on --webpack now. Should have a preliminary PR up soon. There's definitely a ton more work for us to get there compared to --yarn, but I'd like to target 5.1 with that option as well.

**Comment by dhh** [comment]
Not a fan of using components to decompose the Majestic Monolith. But whatever floats your boat! We have a 300-model, 35kloc app in Basecamp 3. Shopify has, I believe, a 700-model 200kloc app. Both Majestic Monoliths, and happily so.

If you, for organizational reasons, want to split the Majestic Monolith, then I think going with completely separate apps is a better way to go.

**Comment by TheLarkInn** [comment]
Feel free and reach out for webpack assistance :) our org would love to track the efforts to adding webpack as a replacement to sprockets etc and assist where valuable.

**Comment by dhh** [comment]
Thanks @TheLarkInn. I'm sure we can use the help! I'll ping you on the PR once it's ready.

The current thinking is to let the asset pipeline focus on JS sprinkles, css, images, etc. I know that web pack technically CAN manage those things, but I don't see a big benefit to doing so.

Where Webpack really shines is for application-level JS. So that's also a way for having the two systems coexist, especially for existing apps. Let for Basecamp, we're considering writing another calendaring feature, which is right in the sweet spot for client-side MVC. We can use Webpack for that feature, which we continue to use the asset pipeline for the JS sprinkles that add sparkle to pages with less demanding UI.

Anyway, I've been diving in over the last several days. Playing with everything within Basecamp. So thinking subject to change!

What's clear is that we want the core Rails principles of convention over configuration to carry over. Nobody should HAVE to configure their own webpack.config.js. It should surely be possible, but not required.

**Comment by TheLarkInn** [comment]
Totalll, I think that's a great start/approach. I'm making a plugin that can potentially work with multipage MVC's [multipage-webpack-plugin](http://github.com/mutualofomaha/multipage-webpack-plugin). It's an oss plugin inspired at work for the need for an easier script injection process for server routes MVC's (we use laravel), so if you find benefit to match this kind of rails arch, let.me know and we can work on it or i can implement additional feature. I'll keep an eye out for an email when you get a PR ready. 👏👏👏

**Comment by moonglum** [comment]
@dhh A different approach would be to offer integration to external asset pipelines, without choosing a specific one like Webpack. I've [built a gem for that purpose](https://github.com/fejo-dk/rails_external_asset_pipeline) and am using that in production with a (Webpack-based) Node.js asset pipeline. The idea is simple:

The external asset pipeline compiles the files. In addition, it creates so called manifest files that map artifact names (like `application.js`) to a path where the file can be found (like `/javascripts/application-gjgkj3499292.js`). The gem simply reads the manifest files so you can still keep using helpers like `stylesheet_link_tag` or `image_tag` without loosing the capability to use cache busting for those files.

If you're interested in that approach, I'm more than happy to contribute that to Rails.

**Comment by dhh** [comment]
Lucas, that's an interesting idea. I'd like to start with Webpack first and see how things go. For starters, we're only going to use Webpack for app-style JS. Not images, css, fonts, etc. I think it's more important out the gate that we nail the experience for the pipelines we provide in the box (Rails' own and Webpack) than going full extendable. But those goals are not in opposition.

I'd be happy to see what an exploration looks like. I'd just temper expectations about what's on a likely path for Rails 5.1.

It'll probably be easier to see how this is going to fit together once we land the current plans for --webpack in master.

**Comment by FND** [comment]
@ddh's approach seems reasonable - however, it might still be advisable to choose a compiler-agnostic name for the CLI option, to prevent API changes when the underlying infrastructure changes? (For straightforward webpack configurations, Rollup can be used as a drop-in replacement.)

**Comment by dhh** [comment]
Don't think the CLI options matter that much, since even if we did go with a generic solution, we can still support that. So --webpack would trigger the Webpack variant of that. Then --other-pipe could trigger that. Or we could extend with --pipeline webpack. This CLI name is only relevant for the start of a new project anyway. So it's not like there'd be backwards incompatibility issues going forward.

But yeah, I think all this is up in the air. Let's get a real PR going, then it'll be easier to talk about specifics.

**Comment by mrreynolds** [comment]
+1 for `--pipeline webpack`.

**Comment by kayakyakr** [comment]
This triggered a long conversation in our Slack about the same and even with just 2 of us mostly talking, we came up with 4 different opinions on how it should be done... In general, I think we'd favor a combination of @moonglum and @dhh's approaches. 

Dunno if anyone's had experience with https://github.com/thoughtbot/ember-cli-rails, but they've mostly hit the sweet spot as far as I'm concerned.

**Comment by dhh** [comment]
@kayakyakr That's always been a big part of the value with Rails. There's an infinite number of ways things COULD be done, but there's an immense benefit to simply picking a path of running with it. That's the core promise behind Convention over Configuration.

**Comment by FND** [comment]
FWIW, without wanting to belabor this point, just a quick clarification: The benefits of Convention over Configuration are of course undisputed - the reason this issue seemed worth raising is because the JavaScript precompiler shouldn't surface to application developers, i.e. the respective source code should be unaffected by whatever is decided here.

**Comment by kayakyakr** [comment]
@FND that's not going to be the case if the implementation is to run sprockets and webpack side-by-side, though. Each pipeline wants its own assets path. Only taking the path of full pipeline replacement gets you the ability to be precompiler-agnostic.

**Comment by moonglum** [comment]
> In general, I think we'd favor a combination of @moonglum and @dhh's approaches.

I think that makes sense. If the ability to read manifest files as described above would be built-into Rails, then all the `--webpack` flag would need to do is create a webpack config that creates the according manifest files. Adding a `--whatever-other-solution-you-need` flag would then be simple.

This would offer the easy out of the box experience that @dhh described, but would still be flexible to use with other external asset pipelines.

**Comment by NekR** [comment]
> I know that web pack technically CAN manage those things, but I don't see a big benefit to doing so.

It might be useful if someone would, say, use a ServiceWorker webpack plugin to generate it for static assets. I know there is ServiceWorker plugin for rails' itself, but still, that might be possible.

**Comment by steakknife** [comment]
Looks good. Previously there was **[webpack-rails](https://github.com/mipearson/webpack-rails)** (more resources: [*Migrating from Rails’ asset pipeline to Node’s webpack*](https://medium.com/@joaomilho/migrating-from-rails-asset-pipeline-to-node-s-webpack-684230e3a93a#.gutq8v28b),  [react-webpack-rails-tutorial](https://github.com/shakacode/react-webpack-rails-tutorial)), which might serve as an example to build upon configuration / conventions and integration behaviors.

**Comment by akabekobeko** [comment]
`webpack` is popular and multi-functional. However, it is mere module and it is not standard as `Node` or `npm`. As with `yarn`, these are **optional** as `npm` and `npm-scripts` as standard.

If [jQuery dependence is to be stopped](https://github.com/rails/rails/issues/25208), it is better to be cautious about non-standard product dependence.

If future `yarn` and `webpack` are obsolete (...It may be impossible), `npm` and `npm-scripts` will remain.

**Comment by justin808** [comment]
@steakknife Thanks for the shout-out for the https://github.com/shakacode/react-webpack-rails-tutorial. It's sibling package is https://github.com/shakacode/react_on_rails.

**Comment by justin808** [comment]
@dhh wrote:
> Lucas, that's an interesting idea. I'd like to start with Webpack first and see how things go. For starters, we're only going to use Webpack for app-style JS. Not images, css, fonts, etc. 

A huge advantage of Webpack is CSS-Modules. If you like tidy, you'll love CSS-Modules. See [Sass, CSS Modules, and Twitter Bootstrap Integration](https://github.com/shakacode/react-webpack-rails-tutorial#sass-css-modules-and-twitter-bootstrap-integration).

Here's an example of it:

* https://github.com/shakacode/react-webpack-rails-tutorial/blob/master/client/app/bundles/comments/components/CommentBox/CommentForm/CommentForm.jsx#L20
* https://github.com/shakacode/react-webpack-rails-tutorial/blob/master/client/app/bundles/comments/components/CommentBox/CommentForm/CommentForm.scss#L3

**Comment by josepjaume** [comment]
@dhh in my experience, the major pain points I've found integrating any sort of pre-compiling using Webpack and feeding it to Sprockets are asset fingerprinting. The way Sprockets works right now, it's nearly impossible to create, let's say, a react component that references a particular asset (an image, for example), without getting React to read Sprocket's manifest file, or Sprockets read Webpack's manifest file.

There's two approaches I believe might work: 

* The first one, **just let Webpack handle asset fingerprinting**, but you already said that's not desirable.

* Another other option would be to let Sprockets fingerprint all the assets and *rewrite* the actual assets at the end, instead of relying on the original code to write the right asset locations - something like what the Phoenix framework does with brunch. That'd would also decouple the contract between webpack (or any other build tool) and sprockets, allowing us to integrate with virtually any other build tool.

Maybe the guys on https://github.com/shakacode/react_on_rails/ (specially @justin808) could chime in on the matter, they seem to have given it a lot of thought.

**Comment by justin808** [comment]
@josepjaume We have the problem solved in React on Rails, such that "it just works". The solution involves:

1. Allow webpack to fingerprint assets.
2. Publish the assets using the Rails asset pipeline, which creates double fingerprints.
3. Create symlinks at deployment, using the manifest file, so that we have single-fingerprinted symlink names that point to the double fingerprinted files.

This way, the correct file names are available on deployment.

Any questions?

**Comment by dhh** [comment]
For those interested, I've published the first work on --webpack here: https://github.com/rails/rails/pull/27288.

Re: cross-reading of manifests, @josepjaume, I think that's actually how we should make this work. Make it possible for each side to read the others manifest and construct what they need from there. The current strategy is to use Webpack exclusively for app-like JS, and nothing else. Asset pipeline for everything else (JS sprinkles, images, fonts, css, etc).

**Comment by iantheparker** [comment]
@josepjaume this is also how the [bower-rails gem](https://github.com/rharriso/bower-rails#relative-asset-paths) resolved relative asset paths.

> Another other option would be to let Sprockets fingerprint all the assets and rewrite the actual assets just at the end of the build process, instead of relying on the original code to write the right asset locations 

Has this option gotten any love since then?

**Comment by krushi123** [comment]
My Rails Version is 5.1.2 and Ruby 2.4 - I am using Yarn

> yarn add font-awesome
Now, in application.css.scss I imported the font-awesome

> @import font-awesome/scss/font-awesome;

Now, when I refer to fa-icon in the view, I am getting 404 Not found error in development. Any idea what I am doing wrong?

**Comment by Schwad** [comment]
Hey all, really interested in what's going on with these changes and updating from Rails 5.0 to Rails 5.1. Quick (hopefully not too stupid) question; does this only apply to new Rails Apps with `rails new blog --yarn`, or are there any out of the box options included for updating existing Rails 5.1 apps with this? 

Thanks so much for your work on this feature!!! 🎉

**Comment by aried3r** [comment]
@Schwad, there is a `rails yarn:install` task, but it doesn't really do everything that happens when a new Rails app is created, notably editing the `.gitignore` file, adding `node_modules` to the assets path etc.

You can find a few things if you [search for `skip_yarn`](https://github.com/rails/rails/search?utf8=%E2%9C%93&q=skip_yarn&type=Code) in the codebase.

I'm not sure if this could be changed or is an intentional omission, but I couldn't find an official guide for adding yarn to existing Rails projects upgrading to 5.1.

**Comment by aried3r** [comment]
@Schwad, also, if you run the `rails app:update` task, Rails will offer you to add all necessary changes except for the `.gitignore` file I think.

**Comment by steakknife** [comment]
I don't use Rails for new development anymore, but yarn and broccoli awesome. Yarn is face-meltingly fast and a nearly drop-in replacement for npm. Webpack proved to be brittle, inflexible and slow, despite having a loyal, zealous herd.

**Comment by amrutjadhav** [comment]
@krushi123 You need to overwrite the `$fa-font-path` variable which is used by font-awesome to point to font directory. Just add these line to your `application.css.scss`:
```$fa-font-path: "font-awesome/fonts/";```
before 
```@import font-awesome/scss/font-awesome;```

**Pull request merged**